File: 1005 Maximize Sum Of Array After K Negations.java
class Solution {
    public int largestSumAfterKNegations(int[] A, int K) {
        PriorityQueue<Integer> pq = new PriorityQueue<Integer>();
        
        for(int x: A) pq.add(x);
        while( K--  > 0) pq.add(-pq.poll());
  
        int sum  = 0;
        for(int i = 0; i < A.length; i++){
            sum += pq.poll();
        }
        return sum;
    }
}

/* Beat 100% by JPV
class Solution {
    public int largestSumAfterKNegations(int[] A, int K) {
        Arrays.sort(A);
        for (int i = 0, j = 0; i < K; i++, A[j] = -A[j]) if (j+1 < A.length && A[j+1] < A[j]) j++;
        int sum = 0;
        for (int i = 0; i < A.length; i++) sum += A[i];
        return sum;
    }
}*/

File: 1005 Maximize Sum Of Array After K Negations.py
class Solution:
    def largestSumAfterKNegations(self, A: List[int], K: int) -> int:
        A.sort()
        # First pass, setting all negatives to positives
        for i in range(len(A)):
            if A[i] >= 0: break
            else: 
                A[i] *= -1
                K -= 1
            if K == 0: return sum(A)
        
        sm = sum(A)
        if K == 0 or K % 2 == 0: return sm
        else: return sm - 2 * min(A)